plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataRebuilder.java

```diff
@@ -0,0 +1,466 @@
+                            metadataUpdater.processMetadata(
+                                tx,
+                                metadataMavenPath(currentGroupId, null, null),
+                                metadataBuilder.onExitGroupId()
+                            );
+                            tx.commit();
+                          }
+                        }
+                      });
+                    }
+                    currentGroupId = groupId;
+                    metadataBuilder.onEnterGroupId(groupId);
+                  }
+                  isolate(new Runnable()
+                  {
+                    @Override
+                    public void run() {
+                      rebuildMetadataInner(groupId, artifactId, baseVersions);
+                    }
+                  });
+                  return true;
+                }
+
+                @Override
+                public void end() {
+                  if (currentGroupId != null) {
+                    isolate(new Runnable()
+                    {
+                      @Override
+                      public void run() {
+                        try (StorageTx tx = storageFacet.openTx()) {
+                          metadataUpdater.processMetadata(
+                              tx,
+                              metadataMavenPath(currentGroupId, null, null),
+                              metadataBuilder.onExitGroupId()
+                          );
+                          tx.commit();
+                        }
+                      }
+                    });
+                  }
+                }
+              }
+          )
+      ).execute(sqlParams);
+    }
+
+    /**
+     * Method rebuilding metadata that performs artifact and baseVersion processing. While it is called from {@link
+     * #rebuildMetadata(ODatabaseDocumentTx)} method, it will use a separate TX/DB to perform writes, it does NOT
+     * accept the TX from caller.
+     */
+    private void rebuildMetadataInner(final String groupId,
+                                      final String artifactId,
+                                      final Set<String> baseVersions)
+    {
+      metadataBuilder.onEnterArtifactId(artifactId);
+      for (String baseVersion : baseVersions) {
+        metadataBuilder.onEnterBaseVersion(baseVersion);
+        try (StorageTx tx = storageFacet.openTx()) {
+          final Iterable<Component> components = tx.findComponents(
+              "group = :groupId and name = :artifactId and attributes.maven2.baseVersion = :baseVersion",
+              ImmutableMap.<String, Object>of(
+                  "groupId", groupId,
+                  "artifactId", artifactId,
+                  "baseVersion", baseVersion
+              ),
+              ImmutableList.of(repository),
+              null // order by
+          );
+          for (Component component : components) {
+            final Iterable<Asset> assets = tx.browseAssets(component);
```

Comment:
Can reduce to `for (Asset asset : tx.browseAssets(component)) {` as assets is not used anywhere else